File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Unique Binary Search Trees.cpp
class Solution
{
public:
    int dp[100010] = {0};
    int numTrees(int n)
    {
        if (n <= 1)
            return 1;
        if (dp[n] != 0)
            return dp[n];
        int ret = 0;
        for (int i = 0; i < n; i++)
            ret += numTrees(i) * numTrees(n - i - 1);
        return dp[n] = ret;
    }
};

File: Online Judges/LeetCode/June LeetCoding Challenge/Week-4/Unique Binary Search Trees.rs
impl Solution {
    pub fn num_trees(n: i32) -> i32 {
        let mut dp = vec![0; (n+1) as usize];
        dp[0] = 1;
        dp[1] = 1;
        for i in 2..n+1{
            for j in 0..i{
                dp[i as usize] = dp[i as usize] + (dp[j as usize] * dp[(i-j-1) as usize]);
            }
        }
        dp[n as usize]
    }
}